### DISCUSSION TITLE
CorrelationId and OperationId in Function Apps

### DISCUSSION THREAD

**Discussion by carolinasantos0** [arcus-azure/arcus.observability#333] [Q&A] [answered]
I want to use the CorrelationId and OperationId in Function App Http trigger. I'm using the latest versions.
First, I've installed Arcus Observability and followed the documentation instructions, but it doesn't show the correlationId nor operationId in application insights. Neither in http response. 

1. Tried to get the ICorrelationInfoAccessor GetCorrelationInfo in function and it was null. (The documentation says to access by "CorrelationInfo", but the method name is showing "GetCorrelationInfo").
 Startup.cs has ` services.AddCorrelation();` `.Enrich.WithCorrelationInfo(new DefaultCorrelationInfoAccessor())` `.WriteTo.AzureApplicationInsights("key")`

2. Tried the Arcus - Web API. Added `.Enrich.WithHttpCorrelationInfo(builder.Services.BuildServiceProvider())` `builder.Services.AddHttpCorrelation(configureOptions);` It worked.
Also the examples in documentation is not very clear.

3. For functions http trigger is recommended to use the Acurs Web API correlation?
4. HttpCorrelation.TryHttpCorrelate should be used when is a custom correlation? What is the advantage?
5. The operationId is generating as "0HMFCMJS4QH8H:00000001" with default configuration. Shouldn't be the same format as TransactionId ?
6. The default names in http correlation response is different from correlation info in Application Insight. Using default configuration.
 Http response: `X-Transaction-ID` and `RequestId`
 Application Insight: `TransactionId` and `OperationId`

**Comment by stijnmoreels** [comment]
Hi 🙋‍♂️, thx for using Arcus.

1-4. First of all, the Azure Functions for which we provide Arcus correlation functionality, is custom bc we still don't have middleware in Azure Functions projects. Your points 1-3 is all about this, so I recommend reading [this section](https://webapi.arcus-azure.net/features/correlation/#installation-1) which describes how one can use the Arcus HTTP correlation inside an Azure Functions HTTP trigger. Also, make sure that you check our [Arcus Template for Azure Functions HTTP trigger](https://templates.arcus-azure.net/azurefunctions-http-template) which not only has built-in HTTP correlation, but also other Arcus recommended functionality.

5. The Operation ID is using the trace identifier in the HTTP request. This is a different identifier than the Transaction ID, as it's uniquely describes an unique operation, instead of describing a group of operations (transaction). See [this part of our documentation](https://observability.arcus-azure.net/Features/correlation) and our [blog post](https://www.codit.eu/blog/announcing-arcus-observability/) about the initial Observability purpose.

6. Yes, that's correct that these names are different. The HTTP correlation default used request header name has also no relation with how Azure Application Insights needs to define its correlation. These are the required names for us to use in order to correctly correlate components in Application Insights. 

We hope that this will guide you a bit further into the Arcus landscape. Let us know if there is anything else we can help you with! And thx for reaching out. 👍

**Comment by fgheysels** [comment, marked as answer]
@carolinasantos0  Can this maybe linked to this issue #336 ?   You can try to remove the call to `ClearProvidersExceptFunctionProviders` and see if it works as expected ?

**Reply by carolinasantos0** [reply]
Yes, I removed and it worked as expected. 
![image](https://user-images.githubusercontent.com/97909792/157645477-b9b64c4c-f397-49d5-b63a-3b0b73aa743a.png)
